diff --git a/drivers/gles3/shaders/scene.glsl b/drivers/gles3/shaders/scene.glsl
index 03515c7..5759d55 100644
--- a/drivers/gles3/shaders/scene.glsl
+++ b/drivers/gles3/shaders/scene.glsl
@@ -148,7 +148,9 @@ layout(std140) uniform DirectionalLightData { //ubo:3
 	mediump vec4 shadow_split_offsets;
 
 	mediump float fade_from;
-	mediump vec3 pad;
+	mediump float pad0;
+	mediump float pad1;
+	mediump float pad2;
 };
 
 #endif //ubershader-skip
@@ -847,7 +849,9 @@ layout(std140) uniform DirectionalLightData {
 	mediump vec4 shadow_split_offsets;
 
 	mediump float fade_from;
-	mediump vec3 pad;
+	mediump float pad0;
+	mediump float pad1;
+	mediump float pad2;
 };
 
 uniform highp sampler2DShadow directional_shadow; // texunit:-5
