diff --git a/core/global_constants.cpp b/core/global_constants.cpp
index f6c15d001f..349e9680e7 100644
--- a/core/global_constants.cpp
+++ b/core/global_constants.cpp
@@ -203,6 +203,8 @@ void register_global_constants() {
 	BIND_GLOBAL_ENUM_CONSTANT(KEY_HELP);
 	BIND_GLOBAL_ENUM_CONSTANT(KEY_DIRECTION_L);
 	BIND_GLOBAL_ENUM_CONSTANT(KEY_DIRECTION_R);
+	BIND_GLOBAL_ENUM_CONSTANT(KEY_BRIGHTNESSDOWN);
+	BIND_GLOBAL_ENUM_CONSTANT(KEY_BRIGHTNESSUP);
 	BIND_GLOBAL_ENUM_CONSTANT(KEY_BACK);
 	BIND_GLOBAL_ENUM_CONSTANT(KEY_FORWARD);
 	BIND_GLOBAL_ENUM_CONSTANT(KEY_STOP);
diff --git a/core/os/keyboard.cpp b/core/os/keyboard.cpp
index 7293fba5e0..21c878dbb5 100644
--- a/core/os/keyboard.cpp
+++ b/core/os/keyboard.cpp
@@ -110,6 +110,8 @@ static const _KeyCodeText _keycodes[] = {
 		{KEY_HELP                          ,"Help"},
 		{KEY_DIRECTION_L                   ,"Direction L"},
 		{KEY_DIRECTION_R                   ,"Direction R"},
+		{KEY_BRIGHTNESSDOWN                ,"BrightnessDown"},
+		{KEY_BRIGHTNESSUP                  ,"BrightnessUp"},
 		{KEY_BACK                          ,"Back"},
 		{KEY_FORWARD                       ,"Forward"},
 		{KEY_STOP                          ,"Stop"},
@@ -345,6 +347,8 @@ bool keycode_has_unicode(uint32_t p_keycode) {
 		case KEY_HELP:
 		case KEY_DIRECTION_L:
 		case KEY_DIRECTION_R:
+		case KEY_BRIGHTNESSDOWN:
+		case KEY_BRIGHTNESSUP:
 		case KEY_BACK:
 		case KEY_FORWARD:
 		case KEY_STOP:
diff --git a/core/os/keyboard.h b/core/os/keyboard.h
index 20863a3c06..8decf8a40c 100644
--- a/core/os/keyboard.h
+++ b/core/os/keyboard.h
@@ -113,6 +113,8 @@ enum KeyList {
 	KEY_HELP = SPKEY | 0x31,
 	KEY_DIRECTION_L = SPKEY | 0x32,
 	KEY_DIRECTION_R = SPKEY | 0x33,
+	KEY_BRIGHTNESSDOWN = SPKEY | 0x34,
+	KEY_BRIGHTNESSUP = SPKEY | 0x35,
 	KEY_BACK = SPKEY | 0x40,
 	KEY_FORWARD = SPKEY | 0x41,
 	KEY_STOP = SPKEY | 0x42,
diff --git a/doc/classes/@GlobalScope.xml b/doc/classes/@GlobalScope.xml
index 198b711198..4a893a2e98 100644
--- a/doc/classes/@GlobalScope.xml
+++ b/doc/classes/@GlobalScope.xml
@@ -346,6 +346,12 @@
 		<constant name="KEY_DIRECTION_R" value="16777267" enum="KeyList">
 			Right Direction key.
 		</constant>
+		<constant name="KEY_BRIGHTNESSDOWN" value="16777268" enum="KeyList">
+			Decreases the screen brightness.
+		</constant>
+		<constant name="KEY_BRIGHTNESSUP" value="16777269" enum="KeyList">
+			Increases the screen brightness.
+		</constant>
 		<constant name="KEY_BACK" value="16777280" enum="KeyList">
 			Media back key. Not to be confused with the Back button on an Android device.
 		</constant>
diff --git a/platform/x11/key_mapping_x11.cpp b/platform/x11/key_mapping_x11.cpp
index 7cb5218580..2a74d27828 100644
--- a/platform/x11/key_mapping_x11.cpp
+++ b/platform/x11/key_mapping_x11.cpp
@@ -149,6 +149,8 @@ static _XTranslatePair _xkeysym_to_keycode[] = {
 	{ XF86XK_AudioLowerVolume, KEY_VOLUMEDOWN },
 	{ XF86XK_AudioMute, KEY_VOLUMEMUTE },
 	{ XF86XK_AudioRaiseVolume, KEY_VOLUMEUP },
+	{ XF86XK_MonBrightnessDown, KEY_BRIGHTNESSDOWN },
+	{ XF86XK_MonBrightnessUp, KEY_BRIGHTNESSUP },
 	{ XF86XK_AudioPlay, KEY_MEDIAPLAY },
 	{ XF86XK_AudioStop, KEY_MEDIASTOP },
 	{ XF86XK_AudioPrev, KEY_MEDIAPREVIOUS },
